### PR TITLE
add proper thread cleanup for active log watching

### PR DESCRIPTION
**Pull request by knifewine**
depends on ccm changes here https://github.com/pcmanus/ccm/pull/515

this makes sure that in cleanup the log watching thread will be joined and allowed to stop.

### CODE REVIEW

**Comment by knifewine** [comment]
upgrade test trial here: http://cassci.datastax.com/view/Upgrades/job/upgrade_tests-all-custom_branch_runs/28/
dtest trial here: http://cassci.datastax.com/view/Dev/view/knifewine/job/trunk_dtest_with_log_watching_cs502/1/

**Comment by knifewine** [comment]
links above updated for newer jobs (upgrade failed mysteriously on first try, dtest had a little bug I fixed).

File: dtest.py [hunk 1]
```diff
@@ -420,12 +421,29 @@ def begin_active_log_watch(self):
         # log watching happens in another thread, but we want it to halt the main
         # thread's execution, which we have to do by registering a signal handler
         signal.signal(signal.SIGINT, self._catch_interrupt)
-        self.cluster.actively_watch_logs_for_error(self._log_error_handler, interval=0.25)
+        # actively_watch_logs_for_error returns an event that needs to be .set() on exit to allow the thread to stop
```
Comment (mambocab) [inline dtest.py]: `.set` isn't called anywhere -- maybe `.join`?

File: dtest.py [hunk 2]
```diff
@@ -420,12 +421,29 @@ def begin_active_log_watch(self):
...
+        Should not be called if begin_active_log_watch was never called to start log watching.
+        Can be called multiple times without error.
+        If not called, threads will remain running until the parent process exits.
+        """
+        try:
+            self._log_watch_thread.join()
+        except AttributeError:
+            # no thread apparently, so nothing to stop
+            # may happen when allow_log_errors is True, since in that case we don't init a logging thread
+            pass
```
Comment (mambocab) [inline dtest.py]: It'd be nice if this signaled whether a log watch thread was actually stopped or not. Maybe return `True` if the `try` block completes, and `False` in the `except` block? I'm not sure what the best API is, but I'm assuming raising a new error isn't what we want to do, if that's a common case.

File: dtest.py [hunk 3]
```diff
@@ -305,6 +305,7 @@ def _cleanup_cluster(self):
                 self.cluster.stop(gently=True)
 
             # Cleanup everything:
+            self.stop_active_log_watch()
```
Comment (mambocab) [inline dtest.py]: How certain are we that `.join` won't raise errors that prevent `self.cluster.remove` from being called below?

**Comment by mambocab** [comment]
Looks reasonable, though I want to talk out my comments before I give a +1, as I'm not very familiar with how this behavior works. I'd be more comfortable if @ptnapoleon had a look as well.

File: dtest.py (hunk 1 above)
Comment (knifewine) [inline dtest.py]: oh right, no longer a relevant comment. that happens internally to the ccm thread code now. will remove and update related docstring.

File: dtest.py (hunk 2 above)
Comment (knifewine) [inline dtest.py]: Yeah, I don't think another error is good in that case. I expect it to be a rare one, but didn't want to halt a test if the thread doesn't exist and throw a wrench into other activities. True/False seems reasonable, will push up a fix.

Comment (knifewine) [inline dtest.py]: pushed. I also separated the join, because if that could somehow raise an attribute error (unsure if it woulcd) but wouldn't want that to be caught unintentionally. and added a reasonable timeout for join to take place, since we don't want to hang things up indefinitely.

File: dtest.py (hunk 3 above)
Comment (knifewine) [inline dtest.py]: hmm, this is a tricky one. I want the watcher to wrap right after stop, in case there's any final shutdown logs written with errors, so we get a final scan of the written log. I also want things to basically blow up so we fix the problem quickly, but don't want to leave clusters laying around.

Comment (knifewine) [inline dtest.py]: maybe I try/except, briefly defer if an exception arises and raise after cluster is properly removed?

Comment (mambocab) [inline dtest.py]: Will `finally:` do the trick?

File: dtest.py (hunk 2 above)
Comment (mambocab) [inline dtest.py]: Looks great.

File: dtest.py (hunk 3 above)
Comment (knifewine) [inline dtest.py]: ayup, much better idea than what I was going for. patched and pushed.

**Comment by knifewine** [comment]
depends on ccm pr https://github.com/pcmanus/ccm/pull/515

**Comment by knifewine** [comment]
now that new errors can be ignored via https://github.com/pcmanus/ccm/pull/518 and https://github.com/riptano/cassandra-dtest/pull/1056 i've kicked off a new upgrade trial job to test out active log watching against this dtest code and the ccm changes it depends on, over here:

http://cassci.datastax.com/view/Upgrades/job/upgrade_tests-all-custom_branch_runs/31/

**Comment by knifewine** [comment]
actually, make that http://cassci.datastax.com/view/Upgrades/job/upgrade_tests-all-custom_branch_runs/32/
since the job flaked once again on openstack

**Comment by knifewine** [comment]
@ptnapoleon ping

**Comment by ptnapoleon** [comment]
Still afraid of the failing job

File: dtest.py [hunk 4]
```diff
@@ -316,13 +316,18 @@ def modify_log(self, cluster):
         self.var_debug(cluster)
         self.var_trace(cluster)
 
+    def _cleanup_last_test_dir(self):
```
Comment (ptnapoleon) [inline dtest.py]: Jim would say we should make this a function and not a method.

File: dtest.py [hunk 5]
```diff
@@ -331,22 +336,24 @@ def _cleanup_cluster(self):
                 self.cluster.stop(gently=True)
 
             # Cleanup everything:
-            debug("removing ccm cluster " + self.cluster.name + " at: " + self.test_path)
-            self.cluster.remove()
+            try:
+                self.stop_active_log_watch()
+            finally:
+                debug("removing ccm cluster " + self.cluster.name + " at: " + self.test_path)
```
Comment (ptnapoleon) [inline dtest.py]: String interpolation? For stylistic consistency, if naught else

File: dtest.py [hunk 6]
```diff
@@ -316,13 +316,18 @@ def modify_log(self, cluster):
...
+            os.remove(LAST_TEST_DIR)
+
     def _cleanup_cluster(self):
         if SILENCE_DRIVER_ON_SHUTDOWN:
             # driver logging is very verbose when nodes start going down -- bump up the level
             logging.getLogger('cassandra').setLevel(logging.CRITICAL)
 
         if KEEP_TEST_DIR:
             self.cluster.stop(gently=RECORD_COVERAGE)
+            self._cleanup_last_test_dir()
```
Comment (ptnapoleon) [inline dtest.py]: Why do we do this here now? Not that it's bad, it just seems orthogonal to the PR and I am not following

File: dtest.py [hunk 7]
```diff
@@ -442,16 +449,39 @@ def begin_active_log_watch(self):
...
-        self.cluster.actively_watch_logs_for_error(self._log_error_handler, interval=0.25)
+        self._log_watch_thread = self.cluster.actively_watch_logs_for_error(self._log_error_handler, interval=0.25)
+
+    def stop_active_log_watch(self):
+        """
+        Joins the log watching thread, which will then exit.
+        Should be called after each test, ideally after nodes are stopped but before cluster files are removed.
+
+        Can be called multiple times without error.
+        If not called, log watchin thread will remain running until the parent process exits.
```
Comment (ptnapoleon) [inline dtest.py]: nit: `watching`

File: dtest.py (hunk 6 above)
Comment (knifewine) [inline dtest.py]: good catch, I don't think that should be there at all. removed and pushed.

File: dtest.py (hunk 4 above)
Comment (knifewine) [inline dtest.py]: yeah, though it doesn't rely on self, it still feels conceptually like it should belong with the other class methods to me. don't have a super strong opinion about it. @mambocab ?

File: dtest.py (hunk 5 above)
Comment (knifewine) [inline dtest.py]: pushed.

File: dtest.py (hunk 7 above)
Comment (knifewine) [inline dtest.py]: thanks, pushed.

**Comment by knifewine** [comment]
all but one thing addressed, I believe.

**Comment by knifewine** [comment]
(though still waiting on my side for a decent run of the upgrade tests against this).

File: dtest.py (hunk 4 above)
Comment (mambocab) [inline dtest.py]: shrug emojus

Comment (mambocab) [inline dtest.py]: I have no idea what `LAST_TEST_DIR` actually does, so I can't speak to what it's like conceptually. Y'all do what you feel.

Comment (ptnapoleon) [inline dtest.py]: Leave as method then

**Comment by knifewine** [comment]
got a clean run of the upgrade job trial here: http://cassci.datastax.com/view/Upgrades/job/upgrade_tests-all-custom_branch_runs/41/testReport/

I think we're ready to merge. @ptnapoleon @mambocab  ?

**Comment by ptnapoleon** [comment]
I agree, +1
